### PR TITLE
AuthN: add scheduler to re-authenticate periodically.

### PR DESCRIPTION
**Pull request by jtk54**
AuthN: add scheduler to re-authenticate periodically. @ttomsu @anotherchrisberry PTAL.

### CODE REVIEW

**Comment by anotherchrisberry** [comment]
If the authentication gets invalidated, this will force a browser redirect on the user, which might be jarring - especially if they're in the middle of something.

In that case, we should pop up a notification to the user that their session has been invalidated, then let them reload the page when they're ready. Something like (in `authenticationInitializer`):
```javascript
...
if (status === 401 && redirect) {
  if (hasAuthenticated) { // add a hasAuthenticated variable above the authenticateUser method and set it to true when the user first authenticates
    notifierService.publish(  // inject notifierService above
          `You have been logged out.
            <a role="button" class="action" onclick="document.location.reload()">Log in</button>`);
  } else {
    var callback = encodeURIComponent($location.absUrl());
...
```

**Comment by jtk54** [comment]
I like this, but one question: if a user is 90% through editing a pipeline and they are prompted to refresh, can they finish making the edits and have them sync? If the authentication expires, the calls to save the pipeline edits won't go through since the user isn't authenticated.

Can we somehow save the state of the edits and reload them after the redirection?

**Comment by anotherchrisberry** [comment]
We *probably* can. For pipelines, it wouldn't be too bad. For modals (e.g. editing a server group configuration)...it would be tough, since we can't deep link into those.

Is there some sort of keep-alive behavior in Gate now? What scenarios will cause authentication to expire?

**Comment by jtk54** [comment]
There's no session keep-alive behavior currently in Gate. Currently the authentication expires after a timeout of 20 minutes, but we can extend the session timeout to be whatever we want (most likely something like 8 hours). We will most likely go this route to make sessions easier to deal with, but still would like to preserve the in-flight changes so that nobody loses the work on a long pipeline because of re-auth.

**Comment by ajordens** [comment]
What if we proactively alerted the user that their session was about to expire ... but with sufficient lead time (5-10min?) that they could finish any inflight operation.

It sounds unnecessarily complicated (_to me_) to support interrupts across the board for modals/pipeline edits, etc.

I do agree with having a longer session timeout and likely some ability within gate to refresh the session (ie. get new groups, assert that the user is still employed, etc.).

**Comment by jtk54** [comment]
Yeah that's better, however we'll have to make an assumption about the default session time. Also then the notification function would have to log out before we re-login.

**Comment by jtk54** [comment]
@ajordens In order to do the `/logout`, we need a few changes in Gate to expose that. For now, I went with a notification that has a button to reload when the authentication expires.

File: app/scripts/modules/core/authentication/authentication.module.js [hunk 1]
```diff
@@ -6,14 +6,20 @@ module.exports = angular.module('spinnaker.authentication', [
...
-  .run(function (authenticationInitializer, settings) {
-    if(settings.authEnabled) {
+  .run(function (schedulerFactory, authenticationInitializer, settings) {
+    if (settings.authEnabled) {
+      // schedule deck to re-authenticate every 10 min.
+      let authenticationScheduler = schedulerFactory.createScheduler(600000);
+      let reauthenticateUser = () => {
+        authenticationInitializer.reauthenticateUser();
+      };
+      let authenticationLoader = authenticationScheduler.subscribe(reauthenticateUser);
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/authentication/authentication.module.js]: You should be able to simplify these four lines to:
```javascript
authenticationScheduler.subscribe(authenticationInitializer.reauthenticateUser)
```
If you want to get really really simplified, collapse these five lines to:
```javascript
schedulerFactory.createScheduler(600000).subscribe(authenticationInitializer.reauthenticateUser);
```

File: app/scripts/modules/core/authentication/authentication.module.js [hunk 2]
```diff
@@ -6,14 +6,20 @@ module.exports = angular.module('spinnaker.authentication', [
   require('./authentication.service.js'),
   require('../config/settings.js'),
   require('./authentication.initializer.service.js'),
-  require('./authentication.interceptor.service.js'),
-  require('./userMenu/userMenu.module.js'),
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/authentication/authentication.module.js]: I don't think you want to remove these dependencies - they're not being brought in anywhere else.

File: app/scripts/modules/core/authentication/authentication.initializer.service.js [hunk 3]
```diff
@@ -22,6 +35,10 @@ module.exports = angular.module('spinnaker.authentication.initializer.service',
         .error(loginRedirect);
     }
 
+    function loginNotification() {
+      notifierService.publish(`You have been logged out. <a role="button" class="action" onclick="document.location.reload(true)">Log in</button>`);
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/authentication/authentication.initializer.service.js]: The `onclick` event should just be:
```javascript
document.location.reload()
```
Passing `true` as a parameter will force the browser to reload everything - JS files, fonts, etc - from the server

File: app/scripts/modules/core/authentication/authentication.module.js (hunk 1 above)
Comment (jtk54) [inline app/scripts/modules/core/authentication/authentication.module.js]: Went for the simplified - fixed.

File: app/scripts/modules/core/authentication/authentication.module.js (hunk 2 above)
Comment (jtk54) [inline app/scripts/modules/core/authentication/authentication.module.js]: Done.

File: app/scripts/modules/core/authentication/authentication.initializer.service.js (hunk 3 above)
Comment (jtk54) [inline app/scripts/modules/core/authentication/authentication.initializer.service.js]: Done

**Comment by anotherchrisberry** [comment]
Does the Gate session expire after 30 minutes _regardless_ of user activity? Or do additional ajax calls the Gate extend the session? Or only the `/auth/info` calls?

**Comment by jtk54** [comment]
Currently, the Gate session expires in 30 minutes regardless of user activity. We are also looking into extending the session in Gate if a user accesses the session within some time frame of expiration.

To sum up what we want:

1. Deck periodically hits `/auth/user`, and if the session expires, notifies the user to re-authenticate.
2. Gate session are extended to 8 hours.
3. Gate extends the user session if it's used within a certain time frame of expiration, otherwise it expires and the user must re-auth via 1.

This PR does 1. 2 is simple enough. And 3 we'll figure out with the `spring-redis-session` bindings.

**Comment by anotherchrisberry** [comment]
@jtk54 what's the timeline on (2)? I'd like to get that in place before pushing this change. Is it just a config value we need to set in a yaml file somewhere?

**Comment by jtk54** [comment]
@anotherchrisberry you can configure the session timeout with `session.expiration` in gate-local.yml. We can change the default also, but I'm not convinced starting with 8 hours by default is the right thing to do. I'd opt more for 1 hour, and have (3) extend the session automatically if you are actively using Deck, since this PR will periodically ping Gate.

**Comment by jtk54** [comment]
> Currently, the Gate session expires in 30 minutes regardless of user activity.

Turns out this statement is false - the session expiration is calculated from time of last use, so the session renewal occurs basically anytime we make a call in the session. Hence (3) above is done.

This PR will ensure that the session is extended by polling `/auth/user`, though I think there are other calls that also poll. Additionally, if the session is somehow cleared (i.e. redis is flushed, session is configured to be less time than the polling calls, etc) this will allow the user to re-auth.

**Comment by anotherchrisberry** [comment]
I feel a lot better about this, knowing that (3) is a non-issue. Feel free to merge when you're ready.

**Pull request merged**